### PR TITLE
Add inside / outside margins and todo outline

### PR DESCRIPTION
**Pull request by Tinggaard**
As discussed in https://github.com/typst/packages/pull/1315, a port of package to allow a note outline, and usage of inside / outside margins.

It does not take text orientation into account, as it is (to my knowledge) not inferrable. I was contemplating implementing it manually, but saw that it broke the logic for drawing the notes overall, so I disregarded it.

I'm not that experienced in writing in Typst yet, so please let me know if there could be improvements :)

### CODE REVIEW

**Comment by ntjess** [comment]
Thanks for the PR! I'm traveling for holidays this week but should be able to review the code afterward, if that timing isn't too late

**Comment by Tinggaard** [comment]
No worries

**Comment by ntjess** [comment]
Thanks for the reminder! It's been a particularly busy week for me, but I will get this reviewed tomorrow.

I appreciate your patience 🙂

**Comment by Tinggaard** [comment]
I was also considering adding additional styling to specify the size of the small upward line, as I'm fond of a smaller one in general. Like the ones shown here:
image

But I'm not 100% sure how the path drawing is achieved though, as it seemed to break the layout when I tried to change the size of the line at line 374
https://github.com/ntjess/typst-drafting/blob/769fbbc2909ad28f346ef8ce4d5c75afd11d4739/drafting.typ#L373-L380

File: drafting.typ [hunk 1]
```diff
@@ -199,38 +205,120 @@
...
+  ): it => {
+    link(it.element.location())[
+      // box with color, followed by content
+      #it.body.children.first() #it.body.children.last()
+      #box(width: 1fr, repeat[.])
+      #it.page
+    ]
+  }
+  // only show if there actually are any todos
+  if show-if-empty or counter(figure.where(kind: "todo")).final().first() > 0 {
```
Comment (ntjess) [inline drafting.typ:292]: Can we instead query for `<margin-note>` and/or `<inline-note>` labels instead of creating figures? Is there a distinct advantage of a hidden figure along with the labels? Should we replace label logic with the figure approach?

File: drafting.typ [hunk 2]
```diff
@@ -19,17 +23,19 @@
   (
     margin-right: none,
     margin-left: none,
+    margin-inside: none,
+    margin-outside: none,
+    page-binding-left: true,
     page-width: none,
     page-offset-x: 0in,
     stroke: red,
+    fill: white,
```
[the remark this answers is not in the document]
Comment (ntjess) [reply drafting.typ:32]: It is especially noticeable on multicolor backgrounds, so I would advocate more strongly for `none` default fill
```suggestion
    fill: none,
```

**Comment by ntjess** [comment]
> I was also considering adding additional styling to specify the size of the small upward line

Something like this? Changing the property `caret-height` to a smaller value like 0.25em will match your image:

```patch
From 64d8c48616a8c74e1ca6e655a2fad1c99f91cd2c Mon Sep 17 00:00:00 2001
From: Nathan Jessurun <ntjessu@gmail.com>
Date: Thu, 12 Dec 2024 20:40:47 -0600
Subject: [PATCH] Allow configurable margin caret height

---
 drafting.typ | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/drafting.typ b/drafting.typ
index a2ba157..96ca03a 100644
--- a/drafting.typ
+++ b/drafting.typ
@@ -33,6 +33,7 @@
     rect: rect,
     side: auto,
     hidden: false,
+    caret-height: 1em,
   ),
 )
 #let note-descent = state("note-descent", (:))
@@ -434,8 +435,8 @@
   let path-pts = _path-from-diffs(
     // make an upward line before coming back down to go all the way to
     // the top of the lettering
-    (0pt, -1em),
-    (0pt, 1em + text-offset),
+    (0pt, -props.caret-height),
+    (0pt, props.caret-height + text-offset),
     (dist-to-margin, 0pt),
     (0pt, dy),
     (1*pct + right-width / 2, 0pt)
@@ -464,8 +465,8 @@
   let text-offset = 0.4em
   let box-width = props.margin-left - 4 * pct
   let path-pts = _path-from-diffs(
-    (0pt, -1em),
-    (0pt, 1em + text-offset),
+    (0pt, -props.caret-height),
+    (0pt, props.caret-height + text-offset),
     (-anchor-x + props.margin-left + 1 * pct, 0pt),
     (-2 * pct, 0pt),
     (0pt, dy),
-- 
2.39.5 (Apple Git-154)
```

We can also have a `margin-trail` dict of props or something, if we want to customize more behaviors down the road

**Comment by Tinggaard** [comment]
Oh my, thanks for a comprehensive review!
I will get to it, but I also have plans for the weekend, so it may take a few days... Just FYI

File: typst.toml [hunk 3]
```diff
@@ -2,11 +2,19 @@
 name = "drafting"
 version = "0.2.1"
 entrypoint = "drafting.typ"
-authors = ["Nathan Jessurun"]
+authors = ["Nathan Jessurun", "Jens Tinggaard <@Tinggaard>"]
```
[the remark this answers is not in the document]
Comment (Tinggaard) [reply typst.toml:5]: Well, I'm not that experienced with Typst - I have only been using it for about 2 months, so I'm not always great at doing things the optimal way (as you see).
So far though I really like writing in Typst, as I actually understand the syntax and logic of it (as opposed to LaTeX), although I'm not as hardcore as you ;)

If you still want me as a maintainer I would be happy to give it a try, but being a "contributor" would also be fine.

File: drafting.typ (hunk 1 above)
Comment (Tinggaard) [reply drafting.typ:310]: I considered that, but to my knowledge, it would not be possible to create the boxes with the color of the todo in the outline if only using labels.

**Comment by Tinggaard** [comment]
> We can also have a `margin-trail` dict of props or something, if we want to customize more behaviors down the road

Not entirely sure what you mean by a margin trail... Is it the path from the location in the text, to the note you wish to customize?

File: drafting.typ [hunk 4]
```diff
@@ -398,7 +513,7 @@
 /// - dy (length): Vertical offset from the note's location -- negative values
 ///   move the note up, positive values move the note down
 /// - ..kwargs (dictionary): Additional properties to apply to the note. Accepted values are keys from `margin-note-defaults`.
-#let margin-note(body, dy: auto, ..kwargs) = {
+#let margin-note(body, dy: auto, ..kwargs) = [ #h(0pt) #{ // h(0pt) forces here().position() to take paragraph indent into account
```
Comment (Tinggaard) [inline drafting.typ:516]: I'm not really happy about this approach btw, if you have a suggestion...

But setting
```typ
#set par(first-line-indent: 1em)
```

Results in notes in a new paragraph not being place properly

File: drafting.typ (hunk 1 above)
Comment (Tinggaard) [reply drafting.typ:312]: Reading this again, we *could* query the labels, but the figure itself still has to persist as far as I can tell.

File: drafting.typ [hunk 5]
```diff
@@ -192,45 +200,145 @@
...
 
+/// get direction of text based on defaults for the language
+/// https://github.com/typst/typst/blob/521ceae889f15f2a93683ab776cd86a423e5dbed/crates/typst-library/src/text/lang.rs#L109
+#let text-direction(dir, lang) = if dir == auto {
+  if lang in (
+    "ar", "dv", "fa", "he", "ks", "pa", "ps", "sd", "ug", "ur", "yi",
+  ) { rtl } else { ltr }
+} else {
+  dir
+}
```
Comment (Tinggaard) [inline drafting.typ:211]: Using
```typ
#set text(dir: rtl)
```
generally breaks the layout.
This also applies to the current (`0.2.1`) version, so maybe it's for another PR...

Comment (Tinggaard) [reply drafting.typ:211]: The function is invoked in a `context` block. I did not add `context` here because I thought it always returned `content` which I've just read is wrong. I can change this function to be `context` aware as well if you'd like.

What I meant by breaking the layout is a general issue though.

File: drafting.typ [hunk 6]
```diff
@@ -447,6 +451,7 @@
   box[
     #place(path(stroke: props.stroke, ..path-pts))
     #place(dx: dist-to-margin + 1 * pct, dy: dy, [#note-rect<margin-note>])
+    #_note-outline-entry(props, body)
```
Comment (reverendpaco) [inline drafting.typ:474]: Hey all ... sorry to butt in here, but this causes a bit of a bug with stateful content.

I've been using the @ntjess repo because it's handling right/left based on inside/outside margins, so I will provide a MRE showing the problem.  

The issue is that line 453 and 454 **both** write `body` (line 453 has `body` as a parameter).  The problem is that  by writing out the same content WHEN that content has a stateful call to something, will result in that stateful code being called twice.  

My minimum reproducible example uses `#cite()` which maintains state about how many citations are made on a page ( the chicago style will write the bibliographic reference as a footnote to the bottom of the page)

So if do something like
```typ
#lorem(10);#margin-note([#lorem(10)  #cite(<codd_relational_1970>) #lorem(20)])#lorem(200)
``` 

I will see the footnote superscript in the margin note.  Good.
image

But the bottom of the page has two references, indicating that the hidden content has statefully
bumped up the citation count.

image

I'm not exactly sure how to solve this, because I don't understand exactly the purpose of 
```typ
  #_note-outline-entry(props, body)
```

but I have commented that out for my particular needs (I am vendoring the file).

**Complete MRE:**
```typ
#import "typst-drafting/drafting.typ": *

#let outside_margin = 2.25in
#let inside_margin = 1in


#set page(
    paper: "a4",
    margin: (
      inside: inside_margin,
      outside: outside_margin,
  )
)
#set-page-properties()
#set-margin-note-defaults(stroke: none)


= This paragraph has a normal cite
#lorem(10);#cite(<codd_relational_1970>);#lorem(40)



#pagebreak()
= This paragraph contains a margin note with a citation that double-counts

#lorem(10);#margin-note([#lorem(10)  #cite(<codd_relational_1970>) #lorem(20)])#lorem(200)

#pagebreak()
#bibliography(style: "chicago-notes", "./data/mre.bib")


////////// mre.bib contents
// @article{codd_relational_1970,
// 	title = {A Relational Model of Data for Large Shared Data Banks},
// 	volume = {13},
// 	doi = {https://doi.org/10.1145/362384.362685},
// 	number = {6},
// 	author = {Codd, E F},
// 	date = {1970},
// 	langid = {english},
// 	keywords = {delightql-cite},
// 	file = {Codd - 1970 - A Relational Model of Data for Large Shared Data B.pdf:/Users/danieleklund/Zotero/storage/QM8TJ6B9/Codd - 1970 - A Relational Model of Data for Large Shared Data B.pdf:application/pdf},
// }

```

**Comment by Tinggaard** [comment]
> I've been using the @ntjess repo because it's handling right/left based on inside/outside margins, so I will provide a MRE showing the problem.

The package does currently not support inside/outside margins afaik, or am I missing something?

---

I see your problem @reverendpaco. The addition was made to allow for an outline entry such that all notes in the document can be listed, along with their content.
Maybe an option would be to make this feature opt-in, such that these figures are not created unless explicitly requested.

**Comment by Tinggaard** [comment]
I will see if I can make it work with labels instead of figures, as @ntjess suggested. This way the content would not be displayed twice.

**Comment by reverendpaco** [comment]
> > I've been using the @ntjess repo because it's handling right/left based on inside/outside margins, so I will provide a MRE showing the problem.
> 
> The package does currently not support inside/outside margins afaik, or am I missing something?
> 


Huh.. that's funny.  You're right.  I have no idea how I got here then 😄 .  It's especially funny because if I bothered to look at what your PR was, I could see that it has nothing to do with inside/outside auto-location.

**Comment by reverendpaco** [comment]
> > > I've been using the @ntjess repo because it's handling right/left based on inside/outside margins, so I will provide a MRE showing the problem.
> > 
> > 
> > The package does currently not support inside/outside margins afaik, or am I missing something?
> 
> Huh.. that's funny. You're right. I have no idea how I got here then 😄 . It's especially funny because if I bothered to look at what your PR was, I could see that it has nothing to do with inside/outside auto-location.

Ok.. Sorry for confusing you and myself.   I DO want to use your repo @Tinggaard .   You are the one who introduces inside/outside margin (which I needed for my tufte-style layout).  I was incorrect to say:

> I've been using the @ntjess repo because it's handling right/left based on inside/outside margins

should have said:

> I've been using the @Tinggaard repo because it's handling right/left based on inside/outside margins

I got your names/repos confused.

File: drafting.typ (hunk 5 above)
Comment (ntjess) [reply drafting.typ:211]: No need to add additional context since it's a private method 👍

**Comment by Tinggaard** [comment]
@reverendpaco I have removed the hidden figure (`#_note-outline-entry(props, body)`), and queried using labels as @ntjess suggested.

I have also tested your provided MRE, and it does not create an additional footnote unless `#note-outline()` is invoked, creating an additional citation for the text in the outline.

**Comment by Tinggaard** [comment]
@ntjess, would you care to review again?

**Comment by ntjess** [comment]
Hey, sorry again for the delay. Yes, I'll be able to review this weekend if it's not too late? Otherwise, I can make time in the next two days.

I want to test a few edge cases which requires me to sit down and make the typst documents

**Comment by Tinggaard** [comment]
Take your time. I just wanted to make sure the PR was not forgotten :)

File: typst.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ntjess) [reply typst.toml:5]: I'd love to have some more help given my lack of availability 😅 

If you're willing to take a look at some new issue requests and push upstream improvements (as you've already done), I'm more than happy to keep you listed

File: tests/misc.typ
Comment (ntjess) [inline tests/misc.typ:1]: @Tinggaard any other good things to test here?

File: drafting.typ [hunk 7]
```diff
@@ -192,53 +217,197 @@
...
 
+#let _get-note-outline-props(note) = {
+  let func = note.func()
+  let defaults = margin-note-defaults.get()
+  // styled types from custom notes have unknown fill/stroke/body
+  // TODO: Maybe a better way to signify "unknown"?
+  let props = (
+    fill: note.at("fill", default: defaults.fill),
+    stroke: note.at("stroke", default: defaults.stroke),
+    body: note.at("body", default: "<body text unkonwn>"),
```
Comment (ntjess) [inline drafting.typ:347]: Styled notes add things to the overview. I think it's fine for now, but we can think of a future PR to grab just the text
image

Comment (ntjess) [reply drafting.typ:347]: @Tinggaard Do you think it's desirable to copy the entire `stroke` instead of just the `paint` for the outline box, which will preserve dashing etc?

File: tests/misc.typ
Comment (Tinggaard) [reply tests/misc.typ:1]: Hmm, that depends on the purpose of the test...
We could basically test all the functionality in `margin-note-defaults`, most important is probably something like `hidden` and `caret-height`.

Also, how is the test evaluated?

File: drafting.typ [hunk 8]
```diff
@@ -192,53 +217,197 @@
...
+      note.location().position(),
+      grid(
+        columns: (1em, 1fr, 10pt),
+        column-gutter: 5pt,
+        align: (top, bottom, bottom),
+        box(
+          fill: note-props.fill,
+          stroke: if note-props.stroke == none {
+            none
+          } else if paint != note-props.fill {
```
Comment (Tinggaard) [inline drafting.typ:378]: Maybe it is of lesser concern (as it seems weird to do), but this panics in the `stroke` constructor if `note-props.stroke` is `none`, which is possible (in the current version).

```typ
#set-margin-note-defaults(stroke: none)
```

**Comment by Tinggaard** [comment]
Regarding some of the ideas for extending the customizability of the library. I would love to have this PR merged first, and then maybe do some development branch based on it (or let main be development, and just make releases when new features are tested through).

What are your thoughts on the development process @ntjess? I would love to have your feedback on the code in general, but if you are busy it could also give a lot of downtime, if we decide for you to review every commit (exaggerating, but I hope you get it).

File: drafting.typ (hunk 7 above)
Comment (ntjess) [reply drafting.typ]: Hmm, I'm hesitant to import another package for 1 function but it may be worthwhile. We can indeed fix it in a future PR if it's relevant.

**Comment by ntjess** [comment]
>  love to have this PR merged first

Agreed!

> development branch

Yep, pretty standard. PRs submitted against the main with small changes that are easy to verify. In the future we can define release documents (like how to update the readme, etc)

> Development process

Definitely I think you're welcome to merge without my approval for things like this to keep it moving. I've been especially delayed with year-end things which made this one harder. My only request would be to have issues submitted before commits, so we can get user / each other's feedback for features before committing them 👍

File: tests/misc.typ
Comment (ntjess) [reply tests/misc.typ:1]: We could, I just mean for this PR specifically, the kind of edge cases it might introduce.

Tests aren't evaluated yet, but other typst packages look for png diffs which we could integrate in the future. For now, I just use eyeballs on the outputs.

**Comment by ntjess** [comment]
Thanks for your patience and the feature @Tinggaard! Do you think it's worth releasing 0.2.2 immediately, or is there anything else to consider first?

**Comment by Tinggaard** [comment]
> Thanks for your patience and the feature @Tinggaard! Do you think it's worth releasing 0.2.2 immediately, or is there anything else to consider first?

I think a release would be appropriate, yeah.
